diff --git a/quic/samples/CMakeLists.txt b/quic/samples/CMakeLists.txt
index 2def302d..ef726308 100644
--- a/quic/samples/CMakeLists.txt
+++ b/quic/samples/CMakeLists.txt
@@ -19,6 +19,12 @@ foreach(header ${QUIC_API_HEADERS_TOINSTALL})
 endforeach()
 
 add_executable(echo echo/main.cpp)
# +add_executable(generic generic/main.cpp
# +        generic/GenericClient.h
# +        generic/GenericHandler.h
# +        generic/GenericServer.h
# +        generic/LogQuicStats.h
# +        )
 
 target_compile_options(
   echo
@@ -26,8 +32,23 @@ target_compile_options(
   ${_QUIC_COMMON_COMPILE_OPTIONS}
 )
 
# +target_compile_options(
# +  generic
# +  PRIVATE
# +  ${_QUIC_COMMON_COMPILE_OPTIONS}
# +)
# +
 target_link_libraries(
   echo PUBLIC
+  gflags
+  mvfst_test_utils
+  ${GFLAGS_LIBRARIES}
+  ${LIBGMOCK_LIBRARIES}
+)
# +
# +target_link_libraries(
# # +  generic PUBLIC
# # +  gflags
#    mvfst_test_utils
#    ${GFLAGS_LIBRARIES}
#    ${LIBGMOCK_LIBRARIES}
